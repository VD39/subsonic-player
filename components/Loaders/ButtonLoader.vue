<template>
  <span>
    <span :class="$style.loader" />
    <span class="visually-hidden">Loading data</span>
  </span>
</template>

<style module>
.loader {
  position: relative;
  display: block;
  width: 15px;
  height: 15px;
  border: 2px solid var(--invert-background-color);
  border-top-color: transparent;
  border-radius: 50%;
  animation: button-loader 0.75s linear infinite;
}

@keyframes button-loader {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}
</style>
