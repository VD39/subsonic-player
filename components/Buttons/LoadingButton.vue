<script setup lang="ts">
import ButtonLoader from '@/components/Loaders/ButtonLoader.vue';

defineProps<{
  disabled?: boolean;
  fullWidth?: boolean;
  loading?: boolean;
}>();
</script>

<template>
  <button
    :class="[
      $style.submitButton,
      {
        [$style.disabled]: disabled,
        [$style.fullWidth]: fullWidth,
      },
    ]"
    :disabled="disabled"
    type="submit"
  >
    <span :class="$style.content">
      <slot />

      <ButtonLoader v-if="loading" :class="$style.loader" />
    </span>
  </button>
</template>

<style module>
.submitButton {
  @mixin align-center;

  position: relative;
  justify-content: center;
  padding: var(--space-8);
  background-color: var(--body-background-color);
  border: 1px solid var(--theme-color);
  border-radius: var(--border-radius-m);

  &:hover {
    background-color: var(--hover-selected-color);
  }
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.fullWidth {
  width: 100%;
}

.content {
  @mixin align-center;

  gap: var(--space-8);
}

.loader {
  margin-left: var(--space-25);
}
</style>
