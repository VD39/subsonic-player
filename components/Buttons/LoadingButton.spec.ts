import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import ButtonLoader from '@/components/Loaders/ButtonLoader.vue';
import LoadingButton from './LoadingButton.vue';

function factory(props = {}) {
  return mount(LoadingButton, {
    props: {
      ...props,
    },
    slots: {
      default: 'Text',
    },
  });
}

describe('IconButton', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when isFullWidth prop is not set', () => {
    it('does not add the fullWidth class', () => {
      expect(wrapper.classes()).not.toContain('fullWidth');
    });
  });

  describe('when isFullWidth prop is set to true', () => {
    beforeEach(() => {
      wrapper = factory({
        isFullWidth: true,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('adds the fullWidth class', () => {
      expect(wrapper.classes()).toContain('fullWidth');
    });
  });

  describe('when isLoading prop is not set', () => {
    it('does not show the ButtonLoader component', () => {
      expect(wrapper.findComponent(ButtonLoader).exists()).toBe(false);
    });
  });

  describe('when isLoading prop is set to true', () => {
    beforeEach(() => {
      wrapper = factory({
        isLoading: true,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the ButtonLoader component', () => {
      expect(wrapper.findComponent(ButtonLoader).exists()).toBe(true);
    });
  });

  describe('when disabled prop is not set', () => {
    it('does not add the disabled class', () => {
      expect(wrapper.classes()).not.toContain('disabled');
    });

    it('sets the correct disabled attribute', async () => {
      expect(wrapper.attributes('disabled')).not.toBeDefined();
    });
  });

  describe('when disabled prop is set as true', () => {
    beforeEach(() => {
      wrapper = factory({
        disabled: true,
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('adds the disabled class', () => {
      expect(wrapper.classes()).toContain('disabled');
    });

    it('sets the correct disabled attribute', async () => {
      expect(wrapper.attributes('disabled')).toBeDefined();
    });
  });
});
