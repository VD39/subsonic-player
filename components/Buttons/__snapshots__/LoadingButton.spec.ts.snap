// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IconButton > matches the snapshot 1`] = `"<button class="submitButton" type="submit"><span class="content">Text<!--v-if--></span></button>"`;

exports[`IconButton > when disabled prop is set as true > matches the snapshot 1`] = `"<button class="submitButton disabled" disabled="" type="submit"><span class="content">Text<!--v-if--></span></button>"`;

exports[`IconButton > when fullWidth prop is set to true > matches the snapshot 1`] = `"<button class="submitButton fullWidth" type="submit"><span class="content">Text<!--v-if--></span></button>"`;

exports[`IconButton > when loading prop is set to true > matches the snapshot 1`] = `"<button class="submitButton" type="submit"><span class="content">Text<span class="loader"><span class="loader"></span><span class="visually-hidden">Loading data</span></span></span></button>"`;
